pypesto/hierarchical/problem.py

```diff
@@ -224,34 +224,44 @@ def inner_problem_from_petab_problem(
         par.ixs = ix_matrices[par.inner_parameter_id]
 
     par_group_types = {
-        tuple(obs_pars.split(';')): {
+        tuple(obs_pars.split(';')): (
             petab_problem.parameter_df.loc[obs_par, PARAMETER_TYPE]
             for obs_par in obs_pars.split(';')
-        }
+        )
         for (obs_id, obs_pars), _ in petab_problem.measurement_df.groupby(
             [petab.OBSERVABLE_ID, petab.OBSERVABLE_PARAMETERS], dropna=True
         )
         if ';' in obs_pars  # prefilter for at least 2 observable parameters
     }
 
     coupled_pars = {
-        par
+        group
         for group, types in par_group_types.items()
         if (
             (InnerParameterType.SCALING in types)
             and (InnerParameterType.OFFSET in types)
         )
-        for par in group
     }
 
+    id_to_par = {par.inner_parameter_id: par for par in inner_parameters}
+
+    # assign coupling
     for par in inner_parameters:
         if par.inner_parameter_type not in [
             InnerParameterType.SCALING,
             InnerParameterType.OFFSET,
         ]:
             continue
-        if par.inner_parameter_id in coupled_pars:
-            par.coupled = True
+        for group in coupled_pars:
+            if par.inner_parameter_id in group:
+                par.coupled = True
+                coupled_parameter_id = group[
+                    group.index(par.inner_parameter_id) - 1
```

Comment:
Yes, it's taking the groups from the `petab.OBSERVABLE_PARMETERS` column of the petab measurement file. And each measurement is checked to have at most only one scaling and at most one offset. Then coupled_pars is checked to have both, so every group will have exactly one offset and one scaling.

This will not work if we introduce some other observable parameters into the hierarchical optimization, but with only scaling and offset it will work nicely.

Reply:
I can make the code nicer if you think this is not clean enough (or exact enough). I saw this as small, sufficient, and correct for the current implementation.

Reply:
I think it's clear enough already, could add a comment, but fine as-is too.